python/pyarrow/_array.pyx

```diff
@@ -992,6 +993,23 @@ cdef class FixedSizeBinaryValue(ArrayValue):
         return cp.PyBytes_FromStringAndSize(data, length)
 
 
+cdef class StructValue(ArrayValue):
+    def as_py(self):
+        cdef:
+            CStructArray* ap
+            vector[shared_ptr[CField]] child_fields = self.type.type.children()
+        ap = <CStructArray*> self.sp_array.get()
+        child_arrays = ap.fields()
+        boxed_arrays = (box_array(child) for child in child_arrays)
+        child_names = (child.get().name() for child in child_fields)
+        # Return the struct as a dict
+        return {
+            name: value for name, value in (
+                (name, child_array[self.index].as_py())
+                for (name, child_array) in
+                izip(child_names, boxed_arrays)
+            ) if value is not None
```

Comment:
Hm, what do you mean by "consistent results with pandas convert"? Are null values not being propagated somewhere? Since this is in `as_py` the value for null is `None`. If there's some inconsistency somewhere regarding null propagation can you open a JIRA about it and cc me?

Regarding memory overhead, `None` is a singleton so each additional value in a collection takes up `sizeof(void*)` additional bytes (as opposed to a copy which would take up a pointer plus the size of the data in the object), which isn't a lot relative to everything else we're storing.

Filtering out `None`s would also return `dict`s that don't match the struct schema, since fields with `None` wouldn't show up in the result, so we definitely need to keep `None`s around.